File: sliding window/java/LongestSubarrayWithOnesAfterReplacement.java
public class LongestSubarrayWithOnesAfterReplacement {

    public static void main(String[] args) {

        System.out.println(getLength(new int [] {0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1} , 2));
        System.out.println(getLength(new int [] {0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1} , 3));
        
    }

    public static int getLength(int [] array, int k) {

        int result = 0;
        int maxOnesInWindow = 0;
        int windowStart = 0;

        for(int windowEnd = 0; windowEnd < array.length; windowEnd++){

            if (array[windowEnd] == 1)
                maxOnesInWindow++;

            if((windowEnd - windowStart + 1) - maxOnesInWindow > k ){

                if(array[windowStart] == 1)
                    maxOnesInWindow--;
                windowStart++;

            }

            result = Math.max(result, windowEnd - windowStart + 1);

        }

        return result;
        
    }
    
}

File: sliding window/typescript/LongestSubarrayWithOnesAfterReplacement.ts
function longestOnes(nums: number[], k: number): number {

    let windowStart = 0;
    let result = 0;
    let max1s = 0;

    for (let windowEnd = 0; windowEnd < nums.length; windowEnd++) {
        if (nums[windowEnd] === 1)
            max1s++;

        if ((windowEnd - windowStart + 1 - max1s) > k) {
            if (nums[windowStart] === 1)
                max1s--;

            windowStart++;
        }

        result = Math.max(result, (windowEnd - windowStart + 1));

    }

    return result;

};

console.log(longestOnes([0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1], 2));
console.log(longestOnes([0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1], 3));